tests/bluetooth/bsim_bt/bsim_test_mesh/src/settings_test_backend.c

```diff
@@ -0,0 +1,218 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "settings_test_backend.h"
+
+#include "kernel.h"
+#include <stdio.h>
+#include <unistd.h>
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <errno.h>
+#include <zephyr.h>
+
+#include <bluetooth/mesh.h>
+
+#define LOG_MODULE_NAME settings_test_backend
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define SETTINGS_FILE "/settings_data.log"
+#define FILE_PATH_LEN (256)
+
+#define ENTRY_LEN_SIZE (4)
+#define ENTRY_NAME_MAX_LEN (SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN)
+#define ENTRY_VAL_MAX_LEN (SETTINGS_MAX_VAL_LEN * 2)
+#define READ_LEN_MAX MAX(ENTRY_VAL_MAX_LEN, ENTRY_NAME_MAX_LEN)
+
+static char backend_file[FILE_PATH_LEN];
+static FILE *fp;
+
+static char len_name_fmt[13];
+static char val_fmt[10];
+static char sk_fmt[7];
+
+struct line_read_ctx {
+	off_t offset;
+	int len;
+};
+
+static int entry_check_and_seek(const char *name)
+{
+	if (fseek(fp, 0, SEEK_SET) < 0) {
+		return -1;
+	}
+
+	char line[READ_LEN_MAX + 1];
+
+	while (fscanf(fp, sk_fmt, line) == 1) {
+		if (strstr(line, name) != NULL) {
+			fgetc(fp);      /* skip newline */
+			return ftell(fp);
+		}
+	};
+
+	return 0;
+}
+
+static ssize_t settings_line_read_cb(void *cb_arg, void *data, size_t len)
+{
+	struct line_read_ctx *valctx = (struct line_read_ctx *)cb_arg;
+	char val[ENTRY_VAL_MAX_LEN + 1];
+
+	if (fseek(fp, valctx->offset, SEEK_SET) < 0 ||
+	    fscanf(fp, val_fmt, val) != 1 ||
+	    (valctx->len / 2) != len) {
+		return 0;
+	}
+
+	uint8_t *n = (uint8_t *) data;
+
+	for (int i = 0; i < len; i++, n++) {
+		if (sscanf(&val[i * 2], "%2hhx", n) != 1) {
+			return 0;
+		};
+	}
+
+	return len;
+}
+
+static int settings_custom_load(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	if (fseek(fp, 0, SEEK_SET) < 0) {
+		return -1;
+	}
+
+	int vallen;
+	char name[ENTRY_NAME_MAX_LEN + 1];
+
+	while (fscanf(fp, len_name_fmt, &vallen, name) == 2) {
+
+		if (vallen <= 0) {
+			continue;
+		}
+
+		/* check for matching subtree */
+		if (arg->subtree != NULL && !strstr(name, arg->subtree)) {
+			continue;
+		}
+
+		LOG_INF("loading entry: %s", name);
+
+		struct line_read_ctx valctx;
+
+		valctx.len = vallen;
+		valctx.offset = ftell(fp);
+		int err = settings_call_set_handler(name, vallen / 2, settings_line_read_cb,
+						    &valctx, arg);
+
+		if (err < 0) {
+			return err;
+		}
+	};
+
+	return 0;
+}
+
+/* Entries are saved to optimize redability of the settings file for test development and
+ * debugging purposes. Format:
+ * <length-hex-str><sapce><entry-key><space><entry-value-hex-str>
+ */
+static int settings_custom_save(struct settings_store *cs, const char *name,
+				const char *value, size_t val_len)
+{
+	if (strlen(name) > ENTRY_NAME_MAX_LEN || val_len > SETTINGS_MAX_VAL_LEN) {
+		return -1;
+	}
+
+	if (entry_check_and_seek(name) == 0) {
+		char bufvname[ENTRY_NAME_MAX_LEN + ENTRY_LEN_SIZE + 3];
+
+		snprintk(bufvname, sizeof(bufvname), "%04x %s ", val_len * 2, name);
```

Comment:
Not actually a problem in the Bluetooth Mesh subsystem, but note that settings names may contain spaces. They cannot contain the character `=`, though, so that might be a nice and readable alternative

Reply:
Since 'sscanf' is used here, string parsing won't work without whitespace. With "=", I will have to fall back to the character-by-character name and value-str parsing in "load".